File: lightbulb-pickers/src/main/java/com/github/RooneyAndShadows/lightbulb/pickers/dialog/DialogDateRangePickerView.java
package com.github.rooneyandshadows.lightbulb.pickers.dialog;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.view.View;

import com.github.rooneyandshadows.java.commons.date.DateUtilsOffsetDate;
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils;
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range.DateRangePickerDialog;
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range.DateRangePickerDialogBuilder;
import com.github.rooneyandshadows.lightbulb.pickers.R;
import com.github.rooneyandshadows.lightbulb.pickers.dialog.base.BaseDialogPickerView;

import java.time.OffsetDateTime;
import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import static com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.*;


@SuppressWarnings({"unused", "UnusedReturnValue"})
public class DialogDateRangePickerView extends BaseDialogPickerView {
    private String datePickerFormat;
    private String datePickerFromText;
    private String datePickerToText;
    private OffsetDateTime[] selection;
    private SelectionChangedListener internalSelectionChangedListeners;
    protected final ArrayList<ValidationCheck> validationCallbacks = new ArrayList<>();
    private final ArrayList<SelectionChangedListener> selectionChangedListeners = new ArrayList<>();

    public void addValidationCheck(ValidationCheck validationCallback) {
        validationCallbacks.add(validationCallback);
    }

    public void addSelectionChangedListener(SelectionChangedListener listener) {
        selectionChangedListeners.add(listener);
    }

    public DialogDateRangePickerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public void setDatePickerFormat(String datePickerFormat) {
        this.datePickerFormat = datePickerFormat;
    }

    public void setDatePickerFromText(String datePickerFromText) {
        this.datePickerFromText = datePickerFromText;
    }

    public void setDatePickerToText(String datePickerToText) {
        this.datePickerToText = datePickerToText;
    }

    public void setPickerRange(OffsetDateTime dateFrom, OffsetDateTime dateTo) {
        if (dateFrom == null || dateTo == null) {
            dateFrom = null;
            dateTo = null;
        } else {
            if (DateUtilsOffsetDate.isDateBefore(dateTo, dateFrom)) {
                OffsetDateTime temp = dateFrom;
                dateFrom = dateTo;
                dateTo = temp;
            }
        }
        if (getDialog() == null) selectInternally(dateFrom, dateTo);
        else getDialog().setSelection(dateFrom, dateTo);
    }

    public OffsetDateTime[] getPickerRange() {
        return selection;
    }

    public boolean hasSelection() {
        return selection != null && selection[0] != null && selection[1] != null;
    }

    @Override
    protected String getViewText() {
        if (selection == null || !hasSelection())
            return ResourceUtils.getPhrase(getContext(), R.string.dialog_date_picker_empty_text);
        String from = DateUtilsOffsetDate.getDateString(datePickerFormat, selection[0]);
        String to = DateUtilsOffsetDate.getDateString(datePickerFormat, selection[1]);
        String viewTextFormat = "{from} - {to}";
        return viewTextFormat.replace("{from}", from).replace("{to}", to);
    }

    @Override
    public boolean validate() {
        boolean isValid = true;
        if (validationEnabled) {
            if (required && !hasSelection()) {
                setErrorEnabled(true);
                setErrorText(pickerRequiredText);
                return false;
            }
            for (ValidationCheck validationCallback : validationCallbacks)
                isValid &= validationCallback.validate(selection);
        }
        if (!isValid)
            setErrorEnabled(true);
        else {
            setErrorEnabled(false);
            setErrorText(null);
        }
        return isValid;
    }

    @Override
    protected void readAttributes(Context context, AttributeSet attrs) {
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.DialogDateRangePickerView, 0, 0);
        try {
            if (selection == null)
                selection = new OffsetDateTime[2];
            datePickerFormat = a.getString(R.styleable.DialogDateRangePickerView_DRPV_DateFormat);
            datePickerFromText = a.getString(R.styleable.DialogDateRangePickerView_DRPV_TextFrom);
            datePickerToText = a.getString(R.styleable.DialogDateRangePickerView_DRPV_TextTo);
            if (datePickerFormat == null || datePickerFormat.equals(""))
                datePickerFormat = "yyyy/MM/dd";
            if (datePickerFromText == null || datePickerFromText.equals(""))
                datePickerFromText = "FROM";
            if (datePickerToText == null || datePickerToText.equals(""))
                datePickerToText = "TO";
        } finally {
            a.recycle();
        }
    }

    @Override
    protected DateRangePickerDialog initializeDialog() {
        return new DateRangePickerDialogBuilder(manager, DIALOG_TAG)
                .withSelection(selection)
                .withCancelOnClickOutsude(true)
                .withTextFrom(datePickerFromText)
                .withTextTo(datePickerToText)
                .withPositiveButton(new DialogButtonConfiguration(pickerDialogPositiveButtonText), (view, dialogFragment) -> updateTextAndValidate())
                .withNegativeButton(new DialogButtonConfiguration(pickerDialogNegativeButtonText), (view, dialogFragment) -> updateTextAndValidate())
                .withOnCancelListener(dialogFragment -> updateTextAndValidate())
                .withOnDateSelectedEvent((oldValue, newValue) -> selectInternally(newValue[0], newValue[1]))
                .withAnimations(pickerDialogAnimationType)
                .buildDialog();
    }

    @Override
    protected DateRangePickerDialog getDialog() {
        return (DateRangePickerDialog) pickerDialog;
    }

    @Override
    protected void dispatchSaveInstanceState(SparseArray<Parcelable> container) {
        dispatchFreezeSelfOnly(container);
    }

    @Override
    protected void dispatchRestoreInstanceState(SparseArray<Parcelable> container) {
        dispatchThawSelfOnly(container);
    }

    @Override
    public Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        SavedState myState = new SavedState(superState);
        myState.selectionFrom = DateUtilsOffsetDate.getDateString(DateUtilsOffsetDate.defaultFormatWithTimeZone, selection[0]);
        myState.selectionTo = DateUtilsOffsetDate.getDateString(DateUtilsOffsetDate.defaultFormatWithTimeZone, selection[1]);
        return myState;
    }

    @Override
    public void onRestoreInstanceState(Parcelable state) {
        SavedState savedState = (SavedState) state;
        OffsetDateTime currentFrom = DateUtilsOffsetDate.getDateFromString(DateUtilsOffsetDate.defaultFormatWithTimeZone, savedState.selectionFrom);
        OffsetDateTime currentTo = DateUtilsOffsetDate.getDateFromString(DateUtilsOffsetDate.defaultFormatWithTimeZone, savedState.selectionTo);
        selection = new OffsetDateTime[]{currentFrom, currentTo};
        super.onRestoreInstanceState(savedState.getSuperState());
    }

    private void selectInternally(OffsetDateTime dateFrom, OffsetDateTime dateTo) {
        selection = new OffsetDateTime[]{dateFrom, dateTo};
        updateTextAndValidate();
        dispatchRangeSelectionChangedEvent();
    }

    private void dispatchRangeSelectionChangedEvent() {
        for (SelectionChangedListener publicChangedListener : selectionChangedListeners)
            publicChangedListener.onSelectionChanged(this, selection);
    }

    private static class SavedState extends View.BaseSavedState {
        private String selectionFrom;
        private String selectionTo;

        SavedState(Parcelable superState) {
            super(superState);
        }

        private SavedState(Parcel in) {
            super(in);
            selectionFrom = in.readString();
            selectionTo = in.readString();
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeString(selectionFrom);
            out.writeString(selectionTo);
        }

        public static final Creator<SavedState> CREATOR
                = new Creator<SavedState>() {
            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

    public interface SelectionChangedListener {
        void onSelectionChanged(DialogDateRangePickerView view, OffsetDateTime[] newRange);
    }

    public interface ValidationCheck {
        boolean validate(OffsetDateTime[] currentSelection);
    }
}

File: lightbulb-pickers/src/main/java/com/github/rooneyandshadows/lightbulb/pickers/dialog/DialogDateRangePickerView.kt
package com.github.rooneyandshadows.lightbulb.pickers.dialog

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import android.util.AttributeSet
import android.util.SparseArray
import androidx.databinding.BindingAdapter
import androidx.databinding.InverseBindingAdapter
import androidx.databinding.InverseBindingListener
import androidx.fragment.app.FragmentManager
import com.github.rooneyandshadows.java.commons.date.DateUtilsOffsetDate
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range.DateRangePickerDialog
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range.DateRangePickerDialog.*
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range.DateRangePickerDialogBuilder
import com.github.rooneyandshadows.lightbulb.pickers.R
import com.github.rooneyandshadows.lightbulb.pickers.dialog.base.BaseDialogPickerView
import java.time.OffsetDateTime
import java.util.*

@Suppress("RedundantOverride", "unused", "UnnecessaryVariable")
class DialogDateRangePickerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : BaseDialogPickerView<DateRange>(context, attrs, defStyleAttr) {
    override val dialog: DateRangePickerDialog
        get() = super.dialog as DateRangePickerDialog
    override val viewText: String
        get() {
            val default = ResourceUtils.getPhrase(context, R.string.dialog_date_picker_empty_text)
            if (!hasSelection) return default
            val format = getDatePickerFormat()
            val from = DateUtilsOffsetDate.getDateString(format, selection!!.from)
            val to = DateUtilsOffsetDate.getDateString(format, selection!!.to)
            val viewTextFormat = "{from} - {to}"
            return viewTextFormat.replace("{from}", from).replace("{to}", to)
        }

    init {
        readAttributes(context, attrs)
    }

    @Override
    override fun getDialogBuilder(
        fragmentManager: FragmentManager,
        dialogTag: String,
    ): BaseDialogBuilder<out BasePickerDialogFragment<DateRange>> {
        return DateRangePickerDialogBuilder(dialogTag, fragmentManager)
    }

    @Override
    override fun dispatchSaveInstanceState(container: SparseArray<Parcelable>) {
        dispatchFreezeSelfOnly(container)
    }

    @Override
    override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>) {
        dispatchThawSelfOnly(container)
    }

    @Override
    override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        val myState = SavedState(superState)
        return myState
    }

    @Override
    override fun onRestoreInstanceState(state: Parcelable) {
        val savedState = state as SavedState
        super.onRestoreInstanceState(savedState.superState)
    }

    fun setRange(start: OffsetDateTime, end: OffsetDateTime) {
        selection = DateRange(start, end)
    }

    fun setDatePickerFormat(format: String?) {
        dialog.setDialogDateFormat(format ?: DEFAULT_DATE_FORMAT)
        updateTextAndValidate()
    }

    fun setDatePickerFromText(text: String?) {
        dialog.setDialogTextFrom(text)
    }

    fun setDatePickerToText(text: String?) {
        dialog.setDialogTextTo(text)
    }

    fun getDatePickerFormat(): String {
        return dialog.dialogDateFormat
    }

    fun getDatePickerFromText(): String? {
        return dialog.dialogTextFrom
    }

    fun getDatePickerToText(): String? {
        return dialog.dialogTextTo
    }

    private fun compareValues(v1: DateRange?, v2: DateRange?): Boolean {
        if ((v1 == null) && (v2 == null)) return true
        if ((v1 == null) || (v2 == null)) return false
        return v1.compare(v2)
    }

    private fun readAttributes(context: Context, attrs: AttributeSet?) {
        val attrTypedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.DialogDateRangePickerView, 0, 0)
        try {
            attrTypedArray.apply {
                getString(R.styleable.DialogDateRangePickerView_drpv_date_format).apply {
                    val default = DEFAULT_DATE_FORMAT
                    whenDialogReady {
                        val dialog = it as DateRangePickerDialog
                        dialog.setDialogDateFormat(if (isNullOrBlank()) default else this)
                    }
                }
                getString(R.styleable.DialogDateRangePickerView_drpv_text_from).apply {
                    val default = ResourceUtils.getPhrase(context, R.string.picker_default_date_from_text)
                    whenDialogReady {
                        val dialog = it as DateRangePickerDialog
                        dialog.setDialogTextFrom(if (isNullOrBlank()) default else this)
                    }
                }
                getString(R.styleable.DialogDateRangePickerView_drpv_text_to).apply {
                    val default = ResourceUtils.getPhrase(context, R.string.picker_default_date_to_text)
                    whenDialogReady {
                        val dialog = it as DateRangePickerDialog
                        dialog.setDialogTextFrom(if (isNullOrBlank()) default else this)
                    }
                }
            }
        } finally {
            attrTypedArray.recycle()
        }
    }

    private class SavedState : BaseSavedState {

        constructor(superState: Parcelable?) : super(superState)

        private constructor(parcel: Parcel) : super(parcel)

        @Override
        override fun writeToParcel(parcel: Parcel, flags: Int) {
            super.writeToParcel(parcel, flags)
        }

        @Override
        override fun describeContents(): Int {
            return 0
        }

        companion object CREATOR : Creator<SavedState> {
            override fun createFromParcel(parcel: Parcel): SavedState {
                return SavedState(parcel)
            }

            override fun newArray(size: Int): Array<SavedState?> {
                return arrayOfNulls(size)
            }
        }
    }

    companion object {
        private const val DEFAULT_DATE_FORMAT = "yyyy/MM/dd"
    }

    object Databinding {
        @BindingAdapter("dateRangePickerSelection")
        @JvmStatic
        fun setDateRange(view: DialogDateRangePickerView, newDateRange: DateRange?) {
            view.selection = newDateRange
        }

        @InverseBindingAdapter(attribute = "dateRangePickerSelection", event = "dateRangeSelectionChanged")
        @JvmStatic
        fun getDateRange(view: DialogDateRangePickerView): DateRange? {
            return view.selection
        }


        @BindingAdapter("dateRangeSelectionChanged")
        @JvmStatic
        fun setListeners(view: DialogDateRangePickerView, attrChange: InverseBindingListener) {
            if (view.hasSelection) attrChange.onChange()
            view.dataBindingListener = SelectionChangedListener { newSelection, oldSelection ->
                if (!view.compareValues(newSelection, oldSelection)) {
                    attrChange.onChange()
                }
            }
        }
    }
}